README
##############################################################################
#                         Makin' Bakon Typing Tutor                          #
#		               version 0.3.2                                 #
#		    Copyright (C) 2002 2003 Stephen Webster                  #
#                                                                            #
#  This program is free software; you can redistribute it and/or modify it   #
#  under the terms of the GNU General Public License as published by the     #
#  Free Software Foundation; either version 2 of the License, or (at your    #
#  option) any later version.                                                #
#                                                                            #
#  This program is distributed in the hope that it will be useful, but       # 
#  WITHOUT ANY WARRANTY; without even the implied warranty of                #
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.                      #
#                                                                            #
#  See the GNU General Public License for more details.                      #
#                                                                            #
#  You should have received a copy of the GNU General Public License along   #
#  with this program; if not, write to the Free Software Foundation, Inc.,   #
#  59 Temple Place, Suite 330, Boston, MA 0.1.2-1307 USA                     #
#                                                                            #
#  Please read the COPYING file for the complete license. The copyright      #
#  holder can be contacted via email: moonunit@ukonline.co.uk                #
##############################################################################

* 1: Warning
* 2: A brief history
* 3: Installation
* 4: Usage
* 5: Command line options
* 6: Bugs and De-bugging
* 7: Getting fortune
* 8: Acknowledgements

1: ------------------------------ WARNING ------------------------------------

Makin' Bakon uses Fortune epigrams for some of its text. The epigrams chosen
are sometimes of a politically or sexually insensative nature. If you are
worried about being offended then please use another typing tutor...

2: ----------------- A very brief history of Makin' Bakon --------------------

The Makin' Bakon Typing Tutor started life, like all worthy ideas, down the
pub. I was looking for a project and a close friend suggested I do a typing
tutor because he didn't think there were any worth using on the Linux/BSD
platforms. (There are, as it goes... Go to http://freshmeat.net/ and do a
search for "typing tutor")

We started talking about other, more famous typing tutors and how dull the
text was. Well, several pints and a few puns later (but before solving the
worlds problems) we hit upon the idea of Makin' Bakon, a typing tutor that
entertains and maybe even offends, as well as doing all the proper typing
tutor stuff.

If you like fortune and the idea of a typing tutor that uses them, and you
need to learn typing to a professional standard, then read on...

3: ---------------------------- Installation ---------------------------------

If you want to, copy the BAKONRC file to .bakonrc in your home directory. If
you don't want then a copy will be created for you when you first start-up
the program. Read the INSTALL file for further instructions.

NOTE: You must have the fortune program and the curses development libs
installed before you can compile Makin' Bakon. 

4: ------------------------------- Usage -------------------------------------

For the seriously impatient, type 'bakon' (minus the quotes, Doh!) and any
command-line options at the prompt and all will be revealed, though I would
ask you to read the following text, it wont take long, honest...

Makin' Bakon is really a game that you play in a terminal screen. You have
to rescue Pig from the clutches of the evil factory farmer by typing as fast
and as accurately as you can. Pig gets hurt if you leave a lesson early and
he gets sliced and diced and made into sausages if you don't reach your
words-per-minute target by the time you quit. Read the online help text for
more information.

As well as the default lessons (which you should do first, they don't take
long), there are five seperate fortune based lessons. Bear in mind that the 
default lessons will improve your typing speed much faster than the fortune
based lessons...

Makin bakon adds the following to your home directory...

~/.bakonrc         -  you can set defaults in this
~/.pigsty/scores   -  holds your usage details
~/.pigsty/epigram  -  temp file holds fortune epigrams

5: ------------------------ Command line options -----------------------------

Color and keyboard type can be defined on the command line, or you can set
defaults in the ~/.bakonrc file. Command line options are as follows: 

-fg (foreground)
-bg (background)
-bd (bold)
-kb (keyboard)

The color choices are black, red, green, yellow, blue, magenta, cyan, white
and trans (transparent). Setting the trans option throughout will leave
terminal settings unchanged, which suits some transparent terminals like
Aterm and Eterm. However, if you choose this path then you may want to set
the terminals BOLD to a different color. See your terminal man page for
instructions.

If you get no colors in Aterm then you probably have the -tn (--termName)
switch set, unset it and all should be well.

You can choose between two keyboard types, American and British: us or uk.

For a complete list of choices, just type "bakon -h". Note that command line
options override ~/.bakonrc defaults.

6: ------------------------- Bugs and De-bugging -----------------------------

a) If something goes horribly wrong and your terminal/console freezes, you
   can kill Makin' Bakon by opening up another terminal/console and typing
   'killall bakon'. You can quit Makin' Bakon at any time simply by typing
   CNTRL-Q, or just Q if in the menu screen.

b) Works fine in the standard Linux consol, Aterm, Eterm and Xterm. Doesn't
   scroll so good in KDE Konsole. If you get no colors in Aterm then you
   probably have the -tn (--termName) switch set, unset it and all should be
   well.

c) This program has been tested with curses 1.96 and 1.97. If you have ANY
   terminal/console problems then check /usr/include/curses.h for your
   version of curses. It's important that you let me know what version you're
   using when sending bug reports. (Also let me know about terminal type or
   $TERM setting in console (do 'echo $TERM' at the prompt). 

d) If the program exits with an exception then let me know what the exception
   was. Also include curses and TERM details as outlined above.

d) If you get a segmentation fault and your terminal isn't responding
   properly then type 'reset' at the prompt. You may not be able to see the
   letters as you type but do it anyway, and then press return.
   
   Send bug reports, with any shell output, curses and term details to the
   adress below.

e) Makin' Bakon can be compiled with the -g flag, which means that you can
   get sensible information out of gdb, the GNU debugger. If you want to do
   this, and you get a segmentation fault OR the program exits with "Aborted"
   then just change the line in src/Makefile that looks like this...
   
   CFLAGS = -O2 -Wall
   
   to look like this...
   
   CFLAGS = -g -Wall

   Once your terminal is behaving properly, enter 'gdb bakon' and press return.
   (gdb is a standard part of most linux/BSD distributions.) At the gdb prompt,
   type 'run' and then use the program as you did before. When you get the seg
   fault, gdb will catch it. Type 'where' or 'bt' at the prompt and cut'n'paste
   the stack trace that follows into a text file.

   Send me the text file, with curses and term details to the address below.
   
   ###########################################################################
   #              Send bug reports to: moonunit@ukonline.co.uk               #
   ###########################################################################

7: --------------------------- Getting Fortune -------------------------------

You will need the Fortune program before you can install Makin' Bakon. It is
usually included as part of your distribution.

If you're running Linux and you haven't got fortune then...

You can get the fortune rpm from:
http://rpmfind.net/linux/RPM/contrib/libc6/i386/fortune-mod-1.2.1-1.i386.html

You'll also need the fortune database rpm:
http://rpmfind.net/linux/RPM/contrib/noarch/noarch/
fortune-mod-fortunes-1.2.1-1.noarch.html

You can install the Offensive rpm if you want: 
http://rpmfind.net/linux/RPM/contrib/noarch/noarch/
fortune-mod-offensive-1.2.1-1.noarch.html

Go to Chris Ausbrooks page at rpmfind.net for the latest versions:
http://rpmfind.net/linux/RPM/Chris_Ausbrooks__weed_bucket.pp.ualr.edu_.html

You can get the tar.gz source (which contains everything you'll need) from:
http://ibiblio.org/pub/Linux/games/amusements/fortune/

If you installed the tar.gz package from ibiblio then you may want to create
a link to the executable if it's not in your PATH. You create the link like
this...
    
    ln -s /usr/local/games/fortune /usr/local/bin/

Or... you can include it in your path by adding a line to your ~/.profile or
~/.bashrc, like this... 
    
    PATH=$PATH":/usr/local/games"

8: -------------------------- Acknowledgements -------------------------------

Thanks to Steve Peck. It was all his idea. He made me do it! He's the real
brains behind this, no really...

Thanks to Xander Soldaat for testing it in KDE. He also made the case for
integrating color.

Thanks to Richard Corby-Siddens for his help with C++ exception handling.

------------------------------------------------------------------------------

Here's a parting epigram...

I think pop music has done more for oral intercourse than anything else 
that has ever happened, and vice versa. -- Frank Zappa

Happy Typing!


CHANGES
Makin' Bakon Typing Tutor * version 0.3.2
Copyright (C) Stephen Webster 2002 2003

--------------------------- release version 0.3.2 ----------------------------
1: Removed ios::in from ofstream declarations in pigsty.cpp, so Makin' Bakon
   now opens scores file properly during start-up. 
2: Updated copyright to include year 2003.
--------------------------- release version 0.3.1 ----------------------------
1: Minor code alterations so that it compiles with GNU GCC 3.2.
--------------------------- release version 0.3.0 ----------------------------
1: Accepts defaults from ~/.bakonrc file for keyboard type, foreground,
   background and bold colors.
2: Accpts command-line options which override rc file defaults.
3: Introduced a new fortune based lesson strictly from offensive archives.
4: Fortune based lessons can now only change lesson stats after lesson_length
   characters have been typed (rather than after every line).
5: Removed preferences utility for keyboard type as this is now set using
   rc file or command-line options.
6: Bonus point system awards one wpm every time target speed is passed.    
7: A lot of the code has been overhauled and the error handling has been
   greatly improved.
8: Default lessons have been divided into groups, effectively halving the
   size fo each lesson. This makes the lessons both shorter and more concise.
--------------------------- release version 0.2.2 ----------------------------
1. Makefile now checks for curses includes.
2. Minor changes to default lesson texts.
--------------------------- release version 0.2.1 ----------------------------
1. Added hand-rolled poetry, startrek and offensive fortune archives.
2. Greatly improved Makefile.
3. Paths to data and fortune archive files are now compile-time defined.
4. Fixed bug that re-paints the virtual keyboard in lesson screens after
   choosing '(R)epeat lesson' or '(N)ext lesson'.
--------------------------- release version 0.2.0 ----------------------------
1: Statistics engine now outputs a running total of the users typing speed
   and accuracy.
2: The lesson screen now includes a virtual keyboard for displaying finger
   placement (layouts reflect keyboard types).
3: [Anorak Alert!] Classes that were almost continuosly in scope, or those
   with a lot of static variables, have now become singletons, thus we no
   longer have to keep track of static initialisations.
4: Wpm score now only gets written to defaults file if it is greater than
   the previous session score.
5: [Anorak Alert!] Makefile now checks for fortune databases before compiling
6: The statistics engine deducts 1 wpm for any number of mistakes in a word,
   where a word is 5 characters.
7: [Anorak Alert!] Fixed bug that caused intermittent core dump in FreeBSD
   (The last element of the main menu wasn't set to NULL, ooops...)
--------------------------- release version 0.1.1 ----------------------------
1: Fixed bug in Makefile so that program compiles properly in FreeBSD.
2: [Anorak Alert!] Moved <time.h> header from stats.cpp to stats.h so that
   program compiles properly in FreeBSD.
------------------------------------------------------------------------------


src/macros.h
/* -- macros.h -- */

/*****************************************************************************
 *                         Makin' Bakon Typing Tutor                         *
 *		               version 0.3.2                                 *
 *		    Copyright (C) 2002 2003 Stephen Webster                  *
 *                                                                           *
 *  This program is free software; you can redistribute it and/or modify it  *
 *  under the terms of the GNU General Public License as published by the    *
 *  Free Software Foundation; either version 2 of the License, or (at your   *
 *  option) any later version.                                               *
 *                                                                           *
 *  This program is distributed in the hope that it will be useful, but      * 
 *  WITHOUT ANY WARRANTY; without even the implied warranty of               *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.                     *
 *                                                                           *
 *  See the GNU General Public License for more details.                     *
 *                                                                           *
 *  You should have received a copy of the GNU General Public License along  *
 *  with this program; if not, write to the Free Software Foundation, Inc.,  *
 *  59 Temple Place, Suite 330, Boston, MA 0.1.2-1307 USA                    *
 *                                                                           *
 *  Please read the COPYING file for the complete license. The copyright     *
 *  holder can be contacted via email: moonunit@ukonline.co.uk               *
 *****************************************************************************/

/*----------------------------------------------------------------------------
 macros.h defines the macros and constatnts for the Makin' Bakon Typing Tutor
 ---------------------------------------------------------------------------*/

#ifndef MACROS_H
#define MACROS_H

#include <iostream>
#include <curses.h>
#include <string>

/*-------------------------------- macros ----------------------------------*/

// get menu array index from lesson number
#define GETINDEX(x) (((x) + 10) / 2)
// get lesson number from menu array index
#define GETLESSON(x) (((x) * 2) - 10) 

/*------------------------------- constants --------------------------------*/

const int PINKY = 1;              // little finger
const int RING = 2;               // ring finger
const int MIDDLE = 3;             // middle finger
const int INDEX = 4;              // index finger
const int LEFT = 0;               // left hand
const int RIGHT = 1;              // right hand

const int KBSIZE = 96;            // sizeof Key array
const int NUMROWS = 5;            // number of rows on a keyboard

const int RSPEED = 4;             // number of speed lessons to search
const int TSPEED = 5;             // total number of speed lessons

const int FMENUSIZE = 10;         // sizeof main menu
const int SMENUSIZE = 20;         // sizeof sub menu
const int MENUSIZE = 30;          // main menu + sub menu
const int CHANGEDEF = 1;          // re-define default lesson
const int LESSONONE = 10;         // first lesson
const int TESTLESSON = 50;        // test lesson
const int LASTLESSON = 49;        // last default lesson
const int NOT_SET = 127;          // variable not set

const short COLOR_TRANS = -1;     // transparent theme

const std::string EMPTY("");                // an empty string
const std::string FORE("fore");             // forground name
const std::string BACK("back");             // background name
const std::string BOLD("bold");             // bold name
const std::string KBRD("keyboard");         // keyboard name
const std::string DEFCOL("trans");          // default color setting
const std::string DEFKBRD("uk");            // default keyboard setting
const std::string USTYPE("us");             // american keyboard type
const std::string UKTYPE("uk");             // british keyboard type

const char D_PIGSTY[] = "/.pigsty";         // /home/<user>/.pigsty/
const char D_SCORES[] = "/scores";          // /home/<user>/.pigsty/scores
const char D_EPIGRAM[] = "/epigram";        // /home/<user>/.pigsty/epigram
const char D_BAKONRC[] = "/.bakonrc";       // /home/<user>/.bakonrc
const char D_BAKONBAK[] = "/BAKONRC";       // <DOC>/BAKONRC

const char PIGSTY[] = ".pigsty";            // /home/<user>/.pigsty/
const char SCORES[] = "scores";             // /home/<user>/.pigsty/scores
const char EPIGRAM[] = "epigram";           // /home/<user>/.pigsty/epigram
const char BAKONRC[] = ".bakonrc";          // /home/<user>/.bakonrc

const char D_WELCOME[] = "/welcome.dat";    // <DATA>/welcome.dat
const char D_HELP[] = "/help.dat";          // <DATA>/help.dat
const char D_PIGTIPS[] = "/pigtips.dat";    // <DATA>/pigtips.dat
const char D_INTROS[] = "/intros.dat";      // <DATA>/intros.dat
const char D_LESSONS[] = "/lessons.dat";    // <DATA>/lessons.dat
const char D_SPEED[] = "/speed.dat";        // <DATA>/speed.dat

// win and lose closing screen strings
const char WIN[] = "Yipeeeeeeeeeeee. I'm a free pig!";
const char LOSE[] = "Mmmmmmmmmmmmmmmmmmmmmmmm. Bacon!";

// command-line option groups
const std::string OPTIONS[]={ "-h", "-kb", "-fg", "-bg", "-bd" };
// size of OPTIONS
const int OSIZE = sizeof(OPTIONS)/sizeof(*OPTIONS);

// curses color-map keys 
const std::string CURSKEYS[]={
    "black", "magenta", "green", "yellow",
    "red", "blue", "cyan", "white", "trans"
};
// size of CURSESKEYS
const int CSIZE = sizeof(CURSKEYS)/sizeof(*CURSKEYS);

// curses color-map values
const short CURSVALS[]={
    COLOR_BLACK, COLOR_MAGENTA, COLOR_GREEN, COLOR_YELLOW,
    COLOR_RED, COLOR_BLUE, COLOR_CYAN, COLOR_WHITE, COLOR_TRANS
};
// size of CURSESVALS
const int DSIZE = sizeof(CURSVALS)/sizeof(*CURSVALS);

// list of ALL default values
const std::string DEFKEYS[]={ FORE, BACK, BOLD, KBRD };
// size of DEFKEYS
const int KSIZE = sizeof(DEFKEYS)/sizeof(*DEFKEYS);

#endif
